File: e2e/api/src/main/kotlin/io/github/carlomicieli/trenako/catalog/db/RailwayRow.kt
/*
 *   Copyright (c) 2022-2023 (C) Carlo Micieli
 *
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 */
package io.github.carlomicieli.trenako.catalog.db

import io.github.carlomicieli.trenako.model.ContactInfo
import io.github.carlomicieli.trenako.model.OrganizationEntityType
import io.github.carlomicieli.trenako.model.RailwayGauge
import io.github.carlomicieli.trenako.model.RailwayPeriodOfActivity
import io.github.carlomicieli.trenako.model.RailwayStatus
import io.github.carlomicieli.trenako.model.RailwayTotalLength
import io.github.carlomicieli.trenako.model.Socials
import io.github.carlomicieli.trenako.model.TrackGauge
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table
import java.math.BigDecimal
import java.time.Instant
import java.time.LocalDate

@Table("railways")
data class RailwayRow(
    @Id
    val railwayId: String,
    val name: String,
    val abbreviation: String?,
    val registeredCompanyName: String?,
    val organizationEntityType: OrganizationEntityType?,
    val descriptionEn: String?,
    val descriptionIt: String?,
    val country: String,
    val operatingSince: LocalDate?,
    val operatingUntil: LocalDate?,
    val status: RailwayStatus?,
    val gaugeMeters: BigDecimal?,
    val trackGauge: TrackGauge?,
    val headquarters: List<String>?,
    val totalLengthMi: BigDecimal?,
    val totalLengthKm: BigDecimal?,
    val contactPhone: String?,
    val contactWebsiteUrl: String?,
    val contactEmail: String?,
    val socialsFacebook: String?,
    val socialsInstagram: String?,
    val socialsLinkedin: String?,
    val socialsTwitter: String?,
    val socialsYoutube: String?,
    val version: Int,
    val createdAt: Instant,
    val lastModifiedAt: Instant? = null
) {
    constructor(
        railwayId: String,
        name: String,
        country: String,
        abbreviation: String? = null,
        registeredCompanyName: String? = null,
        organizationEntityType: OrganizationEntityType? = null,
        description: Map<String, String>? = null,
        railwayPeriodOfActivity: RailwayPeriodOfActivity? = null,
        railwayGauge: RailwayGauge? = null,
        railwayTotalLength: RailwayTotalLength? = null,
        headquarters: List<String>? = null,
        contacts: ContactInfo? = null,
        socials: Socials? = null,
        version: Int = 1,
        createdAt: Instant = Instant.now(),
        lastModifiedAt: Instant? = null
    ) : this(
        railwayId,
        name,
        abbreviation,
        registeredCompanyName,
        organizationEntityType,
        description?.get("en"),
        description?.get("it"),
        country,
        railwayPeriodOfActivity?.operatingSince,
        railwayPeriodOfActivity?.operatingUntil,
        railwayPeriodOfActivity?.status,
        railwayGauge?.meters?.toBigDecimal(),
        railwayGauge?.trackGauge,
        headquarters,
        railwayTotalLength?.kilometers?.toBigDecimal(),
        railwayTotalLength?.miles?.toBigDecimal(),
        contacts?.phone,
        contacts?.websiteUrl?.toString(),
        contacts?.email,
        socials?.facebook?.toString(),
        socials?.instagram?.toString(),
        socials?.linkedin?.toString(),
        socials?.twitter?.toString(),
        socials?.youtube?.toString(),
        version,
        createdAt,
        lastModifiedAt
    )
}

File: crates/libs/data/src/catalog/railways/railway_row.rs
use catalog::common::TrackGauge;
use catalog::railways::period_of_activity::RailwayStatus;
use catalog::railways::railway_id::RailwayId;
use chrono::{DateTime, NaiveDate, Utc};
use common::contacts::{MailAddress, PhoneNumber, WebsiteUrl};
use common::organizations::OrganizationEntityType;
use common::socials::Handler;
use rust_decimal::Decimal;

#[derive(Debug)]
pub struct RailwayRow {
    pub railway_id: RailwayId,
    pub name: String,
    pub abbreviation: Option<String>,
    pub registered_company_name: Option<String>,
    pub organization_entity_type: Option<OrganizationEntityType>,
    pub description_de: Option<String>,
    pub description_en: Option<String>,
    pub description_fr: Option<String>,
    pub description_it: Option<String>,
    pub country: String,
    pub operating_since: Option<NaiveDate>,
    pub operating_until: Option<NaiveDate>,
    pub status: Option<RailwayStatus>,
    pub gauge_meters: Option<Decimal>,
    pub track_gauge: Option<TrackGauge>,
    pub headquarters: Vec<String>,
    pub total_length_mi: Option<Decimal>,
    pub total_length_km: Option<Decimal>,
    pub contact_email: Option<MailAddress>,
    pub contact_website_url: Option<WebsiteUrl>,
    pub contact_phone: Option<PhoneNumber>,
    pub socials_facebook: Option<Handler>,
    pub socials_instagram: Option<Handler>,
    pub socials_linkedin: Option<Handler>,
    pub socials_twitter: Option<Handler>,
    pub socials_youtube: Option<Handler>,
    pub version: i32,
    pub created_at: DateTime<Utc>,
    pub last_modified_at: Option<DateTime<Utc>>,
}

#[cfg(test)]
pub mod test {
    use super::*;
    use isocountry::CountryCode;

    #[allow(dead_code)]
    pub fn new_railway_row(name: &str, country: CountryCode, created_at: DateTime<Utc>) -> RailwayRow {
        RailwayRow {
            railway_id: RailwayId::new(name),
            name: name.to_owned(),
            abbreviation: None,
            registered_company_name: None,
            organization_entity_type: None,
            description_de: None,
            description_en: None,
            description_fr: None,
            description_it: None,
            country: country.alpha2().to_owned(),
            operating_since: None,
            operating_until: None,
            status: None,
            gauge_meters: None,
            track_gauge: None,
            headquarters: Vec::default(),
            total_length_mi: None,
            total_length_km: None,
            contact_email: None,
            contact_website_url: None,
            contact_phone: None,
            socials_facebook: None,
            socials_instagram: None,
            socials_linkedin: None,
            socials_twitter: None,
            socials_youtube: None,
            version: 1,
            created_at,
            last_modified_at: None,
        }
    }
}
